vault/resource_userpass_user.go

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) HashiCorp, Inc.
+// SPDX-License-Identifier: MPL-2.0
+
+package vault
+
+import (
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+
+	"github.com/hashicorp/terraform-provider-vault/internal/provider"
+)
+
+func userpassUserResource() *schema.Resource {
+	fields := map[string]*schema.Schema{
+		"backend": {
+			Type:        schema.TypeString,
+			Optional:    true,
+			Description: "Auth backend to which user will be configured.",
+			ForceNew:    true,
+			Default:     "userpass",
+			// standardise on no beginning or trailing slashes
+			StateFunc: func(v interface{}) string {
+				return strings.Trim(v.(string), "/")
+			},
+		},
+		"username": {
+			Type:         schema.TypeString,
+			Required:     true,
+			ForceNew:     true,
+			Description:  "Username part of userpass.",
+			ValidateFunc: provider.ValidateStringSlug,
+		},
+		"password": {
+			Type:        schema.TypeString,
+			Required:    true,
+			ForceNew:    true,
+			Description: "Password part of userpass.",
+			Sensitive:   true,
+		},
+	}
+
+	addTokenFields(fields, &addTokenFieldsConfig{})
+
+	return &schema.Resource{
+		Create: userpassUserCreate,
+		Read:   ReadWrapper(userpassUserRead),
+		Update: userpassUserUpdate,
+		Delete: userpassUserDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: fields,
+	}
+}
+
+func userPath(backend string, username string) string {
+	return fmt.Sprintf("auth/%s/users/%s", strings.Trim(backend, "/"), strings.Trim(username, "/"))
+}
+
+func usernameFromPath(userId string) string {
```

Comment:
We shouldn't depend on the number of elements in a string since that has caused problems in edge cases where users are using custom paths w/ multiple `/`s. Since we can't say for sure how many `/`s the user may have in their path, it is safer to use regex matching to extract the username and mount. For an example of what this looks like, you can take a look at the [MongoDB Role](https://github.com/hashicorp/terraform-provider-vault/blob/main/vault/resource_mongodbatlas_secret_role.go#L211-L231) resource :)